### PR TITLE
Show "no params (metrics) to display" on compare-runs page when selected runs don't have params (metrics)

### PR DESCRIPTION
**Pull request by WeichenXu123**
Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

## What changes are proposed in this pull request?

Show "no params (metrics) to display" on compare-runs page when selected runs don't have params (metrics)

## How is this patch tested?

Manually

https://user-images.githubusercontent.com/19235986/152780965-f907ecad-ef5d-4f5e-b91c-eb6989a39a5e.mov

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by harupy** [comment]
Could you file the PR on universe?

File: mlflow/server/js/src/experiment-tracking/components/CompareRunView.js [hunk 1]
```diff
@@ -104,17 +103,75 @@ export class CompareRunView extends Component {
...
+      return (
+        <table
+          className='table compare-table compare-run-table'
+          style={{ maxHeight: '500px' }}
+          onScroll={this.onCompareRunTableScrollHandler}
+        >
+          <tbody>{dataRows}</tbody>
+        </table>
+      );
+    }
```
Comment (harupy) [inline mlflow/server/js/src/experiment-tracking/components/CompareRunView.js:125]:
```suggestion
    }
  return (
      <table
        className='table compare-table compare-run-table'
        style={{ maxHeight: '500px' }}
        onScroll={this.onCompareRunTableScrollHandler}
      >
        <tbody>{dataRows}</tbody>
      </table>
    );
```

File: mlflow/server/js/src/experiment-tracking/components/CompareRunView.js [hunk 2]
```diff
@@ -104,17 +103,75 @@ export class CompareRunView extends Component {
...
 
+  renderParamTable(colWidth) {
+    const dataRows = this.renderDataRows(
+      this.props.paramLists,
+      colWidth,
+      this.state.onlyShowParamDiff,
+      true,
+    );
+    if (dataRows.length === 0) {
+      return <h2>No parameters to display.</h2>;
```
Comment (harupy) [inline mlflow/server/js/src/experiment-tracking/components/CompareRunView.js:114]: Can we wrap this with the `FormattedMessage` component for i18n?

File: mlflow/server/js/src/experiment-tracking/components/CompareRunView.js [hunk 3]
```diff
@@ -104,17 +103,75 @@ export class CompareRunView extends Component {
...
+            <i className='fas fa-chart-line' style={{ paddingLeft: '6px' }} />
+          </Link>
+        );
+      },
+      Utils.formatMetric,
+    );
+    if (dataRows.length === 0) {
+      return <h2>No metrics to display.</h2>;
+    } else {
+      return (
```
Comment (harupy) [inline mlflow/server/js/src/experiment-tracking/components/CompareRunView.js:156]: Same here https://github.com/mlflow/mlflow/pull/5352/files#r800715846

File: mlflow/server/js/src/experiment-tracking/components/CompareRunView.js [hunk 4]
```diff
@@ -389,17 +410,13 @@ export class CompareRunView extends Component {
...
-      maxWidth: `${colWidth}px`,
-    };
+    const colWidthStyle = this.genWidthStyle(colWidth);
 
     function isAllEqual(k) {
       return data[k].every((x) => x === data[k][0]);
     }
 
-    return keys.map((k) => {
+    const result = keys.map((k) => {
```
Comment (harupy) [inline mlflow/server/js/src/experiment-tracking/components/CompareRunView.js:419]: Can we refactor the code like this?

```js
    function isAllEqual(arr) {
      return arr.every((x) => x === arr[0]);
    }

    const keys = CompareRunUtil.getKeys(list);
    const data = {};
    keys.forEach((k) => (data[k]['values'] = []));
    list.forEach((records, i) => {
      keys.forEach((k) => data[k].values.push(undefined));
      records.forEach((r) => (data[r.key].values[i] = r.value));
      keys.forEach((k) => (data[k]['allEqual'] = isAllEqual(data[k].values)));
    });

    const colWidthStyle = this.genWidthStyle(colWidth);
    return keys
      .filter((k) => !(onlyShowDiff && data[k].allEqual))
      .map((k) => {
```

File: mlflow/server/js/src/experiment-tracking/components/CompareRunView.js [hunk 5]
```diff
@@ -435,6 +452,8 @@ export class CompareRunView extends Component {
         </tr>
       );
     });
+
+    return result.filter((row) => row !== null);
```
[the remark this answers is not in the document]
Comment (WeichenXu123) [reply mlflow/server/js/src/experiment-tracking/components/CompareRunView.js:456]: any concerns of putting filter at last ? move filter to front seems make code complex

[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/server/js/src/experiment-tracking/components/CompareRunView.js:456]: To me, the current approach seems a bit complicated.

## Existing approach

- Compute `values`
- In `map`,
  - Compute `allEqual`.
  - If `allEqual` is true, return `null`.
- Filter by `null`

## Proposed approach

- Compute `values` and `allEqual`
- Filter by `allEqual`
- Show filtered items

File: mlflow/server/js/src/experiment-tracking/components/CompareRunView.js [hunk 6]
```diff
@@ -104,17 +103,87 @@ export class CompareRunView extends Component {
...
+        );
+      },
+      Utils.formatMetric,
+    );
+    if (dataRows.length === 0) {
+      return (
+        <h2>
+          <FormattedMessage
+            defaultMessage='No metrics to display.'
+            description='No metrics to display.'
```
Comment (harupy) [inline mlflow/server/js/src/experiment-tracking/components/CompareRunView.js:164]: Can we fix the description?

File: mlflow/server/js/src/experiment-tracking/components/CompareRunView.js [hunk 7]
```diff
@@ -104,17 +103,87 @@ export class CompareRunView extends Component {
...
+      colWidth,
+      this.state.onlyShowParamDiff,
+      true,
+    );
+    if (dataRows.length === 0) {
+      return (
+        <h2>
+          <FormattedMessage
+            defaultMessage='No parameters to display.'
+            description='No parameters to display.'
```
Comment (harupy) [inline mlflow/server/js/src/experiment-tracking/components/CompareRunView.js:118]: same here

**Pull request closed without merging**